File: server/users-microservice/src/api/signup/signUpController.py
from flask_restful import Resource
from datetime import datetime
from flask import jsonify, request

from api.users.userSchema import UserSchema
from models.userModel import UserModel
from config.config import trelloMicroserviceUrl

from config.config import jwtSecret
from models.index import db

import requests
import uuid
import json
import jwt

class SignUpController(Resource):
    
    def post(self):
        userToAdd = request.get_json()

        name = userToAdd['name']
        fullname = userToAdd['fullname']
        initials = userToAdd['initials']
        email = userToAdd['email']
        password = userToAdd['password']
        application = userToAdd['application']

        user = UserModel(name, fullname, initials, email, password, application)

        db.session.add(user)
        db.session.commit()

        user_schema = UserSchema(many=False)
        result = user_schema.dump(user)

        payload = {
            'name': name,
            'fullname': fullname,
            'email': email
        }

        headers = {'Content-type': 'application/json; charset=utf-8'}

        trello_response = requests.post(trelloMicroserviceUrl + '/api/v1/signup', data=json.dumps(payload), headers=headers)

        if trello_response.status_code == 200:
            token_identifier = str(uuid.uuid4()) + str(uuid.uuid4());

            payload = {
                'iss': 'users-microservice',
                'sub': 'user_token',
                'csrf': token_identifier,
                'iat': datetime.now(),
                'userName': user.name,
                'userEmail': user.email,
            }

            token = jwt.encode(payload, jwtSecret, 'HS256')

            data =  {
                'csrf': token_identifier
            }

            response = jsonify(data=data)
            response.status_code = 200
            response.set_cookie('jwt', value=token, httponly=True)

            return response
        else:
            return jsonify(error='Something went wrong'), 401

File: server/users-microservice/src/api/signup/signupController.js
'use strict';

import request from 'request';
import Boom from 'boom';

import { trelloMicroserviceUrl } from '../../config/config';
import { generateToken } from '../../utils/tokenController';
import { buildResponse } from '../../utils/responseService';
import { userModel } from '../../models/index';

export const saveUser = (req, res) => {
  userModel.findOne({name: req.body.name})
    .then(user => {
      if (user) {
        throw Boom.create(400, 'That name is already taken');
      } else {
        const user = new userModel({
          name: req.body.name,
          fullname: req.body.fullname,
          initials: req.body.initials,
          email: req.body.email,
          application: req.body.application,
          password: req.body.password
        });

        return user.save();
      }
    })
    .then(user => {
      request.post(
        `${trelloMicroserviceUrl}api/signup`,
        { json: {
          name: user.name,
          fullname: user.fullname,
          email: user.email
        } },
        function (error, res2, body) {
          if (!error && res2.statusCode === 200) {
            let { token, csrf } = generateToken(user.name, user.email, user._id);
        
            res.cookie('jwt', token, { httpOnly: true });
            buildResponse(200, csrf, res);
          } else {
            user
              .remove()
              .then(err => {
                if (err) {
                  buildResponse(500, err, res);
                } else {
                  buildResponse(500, 'The user was not created successfully', res);
                }
              })
          }
        }
      );
    })
    .catch(error => {
      if (error.isBoom) {
        buildResponse(error.output.statusCode, error.message, res);
      } else {
        buildResponse(500, error, res);
      }
    });
}; 